server/java/src/main/java/com/godaddy/asherah/grpc/AppEncryptionImpl.java

```diff
@@ -0,0 +1,158 @@
+package com.godaddy.asherah.grpc;
+
+import java.nio.charset.StandardCharsets;
+
+import com.godaddy.asherah.appencryption.Session;
+import com.godaddy.asherah.appencryption.SessionFactory;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.protobuf.ByteString;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.grpc.stub.StreamObserver;
+
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.SessionRequest;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.SessionResponse;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.DataRowRecord;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.EnvelopeKeyRecord;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.KeyMeta;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.EncryptResponse;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.DecryptResponse;
+import static com.godaddy.asherah.grpc.AppEncryptionProtos.ErrorResponse;
+
+public class AppEncryptionImpl extends AppEncryptionGrpc.AppEncryptionImplBase {
+  private static final Logger logger = LoggerFactory.getLogger(AppEncryptionImpl.class);
+
+  private final SessionFactory sessionFactory;
+
+  public AppEncryptionImpl() {
+    sessionFactory = AppEncryptionServer.getSessionFactory();
+  }
+
+  @Override
+  public StreamObserver<SessionRequest> session(final StreamObserver<SessionResponse> responseObserver) {
+
+    logger.info("Connecting stream observer");
+
+    StreamObserver<SessionRequest> streamObserver = new StreamObserver<SessionRequest>() {
+
+      private Session<byte[], byte[]> sessionBytes;
+      private String partitionId;
+
+      @Override
+      public void onNext(final SessionRequest sessionRequest) {
+        // For debug purposes
+        // System.out.println("sessionRequest = " + sessionRequest);
+
+        if (sessionRequest.hasGetSession()) {
+          // Handle response for get session
+          partitionId = sessionRequest.getGetSession().getPartitionId();
+          logger.info("Attempting to create session for partitionId={}", partitionId);
+          sessionBytes = sessionFactory.getSessionBytes(partitionId);
+          responseObserver.onNext(SessionResponse.getDefaultInstance());
+        }
+
+        if (sessionBytes == null) {
```

Comment:
an accidental re-org of the if blocks might break this flow. think having the null check within the encrypt and decrypt blocks would mitigate that

Reply:
looks like this concern is still active as the code doesn't make it clear that `GetSession` needs to precede `Encrypt\Decrypt` operations.